File: Backend-Node/routes/adminDashboard.js
const express = require('express');
const router = express.Router();
const User = require('../models/User');
const Report = require('../models/Report');

/**
 * @swagger
 * /admin/dashboard/user-growth:
 *   get:
 *     summary: Get user growth statistics by month
 *     tags: [Admin Dashboard]
 *     responses:
 *       200:
 *         description: User growth data retrieved successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 growth:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       month:
 *                         type: string
 *                       count:
 *                         type: number
 */
router.get('/dashboard/user-growth', async (req, res) => {
  try {
    const users = await User.find({});
    const monthly = {};

    for (const user of users) {
      const created = user.created_at || new Date();
      const month = created.toISOString().slice(0, 7); // YYYY-MM format
      monthly[month] = (monthly[month] || 0) + 1;
    }

    // Sort by month and format response
    const sortedMonths = Object.keys(monthly).sort();
    const data = sortedMonths.map(m => ({
      month: m,
      count: monthly[m]
    }));

    res.status(200).json({ growth: data });
  } catch (error) {
    console.error('User growth error:', error);
    res.status(500).json({ error: 'Server error fetching user growth' });
  }
});

/**
 * @swagger
 * /admin/dashboard/recent-activity:
 *   get:
 *     summary: Get recent activity (last 5 uploads)
 *     tags: [Admin Dashboard]
 *     responses:
 *       200:
 *         description: Recent activity retrieved successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 recent:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       user:
 *                         type: string
 *                       action:
 *                         type: string
 *                       time:
 *                         type: string
 */
router.get('/dashboard/recent-activity', async (req, res) => {
  try {
    const reports = await Report.find({})
      .sort({ uploaded_at: -1 })
      .limit(5);

    const result = reports.map(r => ({
      user: r.user_email,
      action: `uploaded ${r.file_name}`,
      time: r.uploaded_at
    }));

    res.status(200).json({ recent: result });
  } catch (error) {
    console.error('Recent activity error:', error);
    res.status(500).json({ error: 'Server error fetching recent activity' });
  }
});

/**
 * @swagger
 * /admin/dashboard/report-status:
 *   get:
 *     summary: Get report status distribution
 *     tags: [Admin Dashboard]
 *     responses:
 *       200:
 *         description: Report status counts retrieved successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 normal:
 *                   type: number
 *                 abnormal:
 *                   type: number
 *                 critical:
 *                   type: number
 */
router.get('/dashboard/report-status', async (req, res) => {
  try {
    const reports = await Report.find({});
    const statusCount = {
      normal: 0,
      abnormal: 0,
      critical: 0
    };

    for (const report of reports) {
      const severity = report.ai_summary?.severity || 'low';

      if (severity === 'low') {
        statusCount.normal++;
      } else if (severity === 'medium') {
        statusCount.abnormal++;
      } else {
        statusCount.critical++;
      }
    }

    res.status(200).json(statusCount);
  } catch (error) {
    console.error('Report status error:', error);
    res.status(500).json({ error: 'Server error fetching report status' });
  }
});

module.exports = router;

File: Backend/routes/admin_dashboard.py
from flask import Blueprint, jsonify
from pymongo import MongoClient
from bson import ObjectId
from datetime import datetime
from collections import Counter

mongo = MongoClient("mongodb://localhost:27017/")
db = mongo["LabInsight"]

users_col = db["users"]
reports_col = db["reports"]

admin_dashboard_bp = Blueprint("admin_dashboard", __name__)


# ------------------------------------------------
# 1️⃣ USER GROWTH (COUNT USERS PER MONTH)
# ------------------------------------------------
@admin_dashboard_bp.route("/dashboard/user-growth", methods=["GET"])
def user_growth():
    users = list(users_col.find({}))

    monthly = {}

    for u in users:
        created = u.get("created_at", datetime.now())
        month = created.strftime("%Y-%m")
        monthly[month] = monthly.get(month, 0) + 1

    # Convert to list format for frontend
    data = [{"month": m, "count": monthly[m]} for m in sorted(monthly.keys())]

    return jsonify({"growth": data})


# ------------------------------------------------
# 2️⃣ RECENT ACTIVITY (LAST 5 REPORT UPLOADS)
# ------------------------------------------------
@admin_dashboard_bp.route("/dashboard/recent-activity", methods=["GET"])
def recent_activity():
    reports = list(reports_col.find().sort("uploaded_at", -1).limit(5))

    result = []
    for r in reports:
        result.append({
            "user": r.get("user_email"),
            "action": f"uploaded {r.get('file_name')}",
            "time": r.get("uploaded_at")
        })

    return jsonify({"recent": result})


# ------------------------------------------------
# 3️⃣ REPORT STATUS COUNTS (normal / abnormal / critical)
# ------------------------------------------------
@admin_dashboard_bp.route("/dashboard/report-status", methods=["GET"])
def report_status():
    reports = list(reports_col.find({}))

    status_count = {"normal": 0, "abnormal": 0, "critical": 0}

    for r in reports:
        sev = r.get("ai_summary", {}).get("severity", "low")

        if sev == "low":
            status_count["normal"] += 1
        elif sev == "medium":
            status_count["abnormal"] += 1
        else:
            status_count["critical"] += 1

    return jsonify(status_count)
